add: Plugin controller documentation (#202)
glpi-project/docdev: source/plugins/controllers.rst

Controllers
-----------

Plugin controllers follow the same principles as :doc:`core controllers </devapi/controllers>` with a few differences specific to the plugin system.

.. note::

   Controllers require GLPI >= 11.0.

Creating a plugin controller
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Requirements:

* The controller file must be placed in the ``src/Controller/`` folder of the plugin.
* The namespace must follow PSR-4: ``GlpiPlugin\MyPlugin\Controller\``.

> * The controller must extend ``Glpi\Controller\AbstractController``.

Reviewer: Currently, plugins controllers must be declared public manually. Maybe we should do it automatically in the `Glpi\DependencyInjection\PluginContainer` class.

```suggestion
* The controller must extend ``Glpi\Controller\AbstractController`` or implement the ``Glpi\\DependencyInjection\\PublicService`` interface.
```

* The controller must define a route using the ``Route`` attribute.

> * The controller must return some kind of response.

Reviewer: ```suggestion
* The controller must return a ``Symfony\\Component\\HttpFoundation\\Response`` instance.
```

Controllers placed in ``src/Controller/`` are **automatically discovered**, no manual registration is needed.

Example for a plugin named ``myplugin``:

.. code-block:: php

   # plugins/myplugin/src/Controller/HelloController.php
   <?php

   namespace GlpiPlugin\Myplugin\Controller;

   use Glpi\Controller\AbstractController;
   use Symfony\Component\HttpFoundation\JsonResponse;
   use Symfony\Component\HttpFoundation\Request;
   use Symfony\Component\HttpFoundation\Response;
   use Symfony\Component\Routing\Attribute\Route;

   final class HelloController extends AbstractController
   {
       #[Route("/Hello", name: "myplugin_hello", methods: "GET")]
       public function __invoke(Request $request): Response
       {
           return new JsonResponse(['message' => 'Hello from myplugin!']);
       }
   }

URL routing
^^^^^^^^^^^

Plugin routes are automatically prefixed with the plugin base path. A route defined as ``/Hello`` in the ``myplugin`` plugin will be accessible at:

* ``/plugins/myplugin/Hello`` (standard plugins directory)

> * ``/marketplace/myplugin/Hello`` (marketplace plugins directory)

Reviewer 2: ```suggestion
```

I think marketplace URLs are depreciated.

Reviewer: They are. The `/plugins/` prefix is not related to the filesystem location of the plugin.

Author: Should I remove it from the doc, as when testing I can access for plugin in both url.

A plugin installed from the marketplace would have /marketplace ? 

i.e: Bot work on my local env `http://localhost:8011/marketplace/glpiai/config`  and `http://localhost:8011/plugins/glpiai/config`

Reviewer: You should remove it. A plugin installed from the marketplace will generate `/plugins/XXX` URLs.

   A bug in GLPI prior to **11.0.7** caused plugin routes with method constraints other than ``GET`` to never match.
   The router context was always evaluated as ``GET``, so any route declared with only ``POST``, ``PUT``, ``DELETE``, ``PATCH``, etc. would never be found.

   This bug was fixed in GLPI 11.0.7.
   If your plugin needs to support GLPI < 11.0.7, use the following workaround:
   include ``GET`` alongside the intended methods and check the actual method manually inside the controller.

**Workaround for GLPI < 11.0.7:**

.. code-block:: php

   <?php
   // ❌ Non-GET method only, broken on GLPI < 11.0.7
   #[Route("/MyAction", name: "myplugin_my_action", methods: ['POST'])]

   // ✅ Works on all versions >= 11.0 (check the method manually if needed)
   #[Route("/MyAction", name: "myplugin_my_action", methods: ['GET', 'POST'])]
   public function __invoke(Request $request): Response
   {
       if (!$request->isMethod('POST')) {
           throw new \Symfony\Component\HttpKernel\Exception\MethodNotAllowedHttpException(['POST']);
       }
       // ...
   }

On GLPI >= 11.0.7, you can safely restrict routes to any HTTP method without the workaround.

Stateless routes

> ^^^^^^^^^^^^^^^^

Reviewer 2: Whats the difference with using the firewall attributes to allow unauthenticated users? It should be explained.

Reviewer: Indeed, the difference is that no session is opened and no session cookie is sent. It has to be used when the session opening must be controled by the controller itself (e.g. an API endpoint).

> By default, GLPI requires an active session to access any route (see :doc:`/devapi/controllers`Firewall section).

Reviewer: By default, it only allows profiles that access to the normal interface. The `SecurityStrategy` can be used to adapt the strategy to use.

       // ...
   }

On GLPI >= 11.0.7, you can safely restrict routes to any HTTP method without the workaround.

Unauthenticated access
^^^^^^^^^^^^^^^^^^^^^^

GLPI offers two distinct mechanisms for routes that must be accessible without a logged-in user.
Choosing the right one depends on whether the route needs a session at all.

Session based: No auth check
++++++++++++++++++++++++++++

The session is started normally (the session cookie is read and written), but no authentication check is performed.

The controller can read the current user's session if one happens to be active, but the request is also accepted from anonymous visitors.

Use this for public web pages (e.g. a public form or a login endpoint).

.. code-block:: php

   <?php
   #[Route("/MyAction", name: "myplugin_my_action", methods: ['GET'])]
   #[Glpi\Security\Attribute\SecurityStrategy(Glpi\Http\Firewall::STRATEGY_NO_CHECK)]
   public function __invoke(Request $request): Response
   {
       // Session may or may not be active so do not assume the user is logged in.
   }

No session: Stateless
+++++++++++++++++++++

No session is started and no session cookie is sent or read. The request is fully stateless.

> Use this when the controller manages its own authentication (i.g., an API endpoint that expects a token in a header).

Reviewer: ```suggestion
Use this when the controller manages its own authentication (e.g. an API endpoint that expects a token in a header).
```

> It also prevents session cookies from interfering with HTTP caching for machine-to-machine endpoints.

Reviewer: IMHO, regarding the issues we had to handle in the past with GLPI administrators trying to do improbable stuffs to enfore endpoints caching, we should not mention caceability in the current doc.

```suggestion
```
